A day in the life (#3225)
quicwg/base-drafts: draft-ietf-quic-transport.md

## Life of a QUIC Connection {#connection-lifecycle}

A QUIC connection is a stateful interaction between a client and server, the
primary purpose of which is to support the exchange of data by an application

> protocol.  Streams ({{streams}}) is the primary mechanism by which an

Reviewer: s/is/are?

Author: I really didn't know which here.  But I think that you are right.

> protocol exchanges information.

Reviewer 2: That is not necessarily the case with DATAGRAM extension.

Reviewer 3: Heh. True, though since that extension is not part of the transport draft and not discussed in it at all, I think this sentence is quite correct as is.

> Each connection starts with an handshake phase, during which client and server

Reviewer: s/an/a

establish a shared secret using the cryptographic handshake protocol
{{QUIC-TLS}} and negotiate the application protocol.  The handshake

> ({{handshake}}) confirms that endpoints are willing to communicate

Reviewer 2: ```suggestion
({{handshake}}) confirms that both endpoints are willing to communicate
```

({{validate-handshake}}) and establishes parameters for the connection
({{transport-parameters}}).

An application protocol can also operate in a limited fashion during the
handshake phase.  0-RTT allows application messages to be sent by a client
before receiving any messages from the server.  However, 0-RTT lacks certain key

> security guarantees. In particular, there is no protection against replay

Reviewer 4: ```suggestion
security guarantees.  In particular, there is no protection against replay
```

attacks in 0-RTT; see {{QUIC-TLS}}.  Separately, a server can also send
application data to a client before it receives the final cryptographic
handshake messages that allow it to confirm the identity and liveness of the
client.  These capabilities allow an application protocol to offer the option to

> trade some security guarantees for improved latency.

Reviewer 2: ```suggestion
trade some security guarantees for reduced latency.
```

... or to buffer packets until the handshake is confirmed.

Reviewer 2: This section reads as if 1-RTT could be relaxed while it probably only intends to address 0-RTT.  See also #3224

The use of connection IDs ({{connection-id}}) allows connections to migrate to a
new network path, both as a direct choice of an endpoint and when forced by a
change in a middlebox.  {{migration}} describes how a migration can be performed

> securely without adversely affecting privacy.

Reviewer 2: That is a strong claim. "... with as little impact on privacy as practically possible" would be more precise.

Author: I'm surprised that you aren't also contesting the security claim as well :)  Because that too is in question.

Reviewer 2: We all know that security is an illusion :)

Reviewer 4: {{migration}} describes potential mitigations for the security and privacy implications of migration.

For connections that are no longer needed or desired, there are several ways for

> a client and server to terminate a connection ({{termination}}).

Reviewer 5: The PR talks about birth, marriage, divorce and death. 

Perhaps it is pertinent to have a paragraph the mentions the connection's role in the steady-state between some of these things. In fact, mentioning that it is mainly the application mapping's responsibility to define the specifics might help.

> a client and server to agree to remove connection state ({{termination}}).

Reviewer: This paragraph feels incomplete. What are those different ways? Are you referring to silent close vs. immediate close?

Author: Yeah, I had that in the first cut, but then removed it.  The referenced section has enough information about closing.  It's not super-important to duplicate all that text when it is so succinctly stated later.

> a client and server to agree to remove connection state ({{termination}}).

Reviewer 6: I'm not sure this suggestion is better, but it reads a bit more easily to me.

```suggestion
a client and server to terminate a connection ({{termination}}).
```

A QUIC connection is a stateful interaction between a client and server, the
primary purpose of which is to support the exchange of data by an application
protocol.  Streams ({{streams}}) are the primary means by which an application
protocol exchanges information.

> Each connection starts with a handshake phase, during which client and server

Reviewer 2: unless of course your intended audience is primarily from East End

Reviewer 2: @MikeBishop Cockney  an 'andshake
